test/test_pebble.py

```diff
@@ -1076,6 +1161,423 @@ def test_get_services_names(self):
+        self.assertEqual(len(self.client.requests), 1)
+        request = self.client.requests[0]
+        self.assertEqual(request[:3], ('POST', '/v1/files', None))
+
+        headers, body = request[3:]
+        content_type = headers['Content-Type']
+        dest = io.BytesIO()
+        req = self._parse_write_multipart(content_type, body, {'/foo/bar': dest})
+        self.assertEqual(dest.getvalue(), b'content')
+        self.assertEqual(req, {
+            'action': 'write',
+            'files': [{'path': '/foo/bar'}],
+        })
+
+    def test_write_file_options(self):
+        self.client.responses.append((
+            {'Content-Type': 'application/json'},
+            b"""
+{
+    "result": [
+        {"path": "/foo/bar"}
+    ],
+    "status": "OK",
+    "status-code": 200,
+    "type": "sync"
+}
+""",
+        ))
+
+        source = io.BytesIO(b'content')
+        self.client.write_file('/foo/bar', source, make_dirs=True, permissions=0o600,
+                               user='bob', group='staff')
+
+        self.assertEqual(len(self.client.requests), 1)
+        request = self.client.requests[0]
+        self.assertEqual(request[:3], ('POST', '/v1/files', None))
+
+        headers, body = request[3:]
+        content_type = headers['Content-Type']
+        dest = io.BytesIO()
+        req = self._parse_write_multipart(content_type, body, {'/foo/bar': dest})
+        self.assertEqual(dest.getvalue(), b'content')
+        self.assertEqual(req, {
+            'action': 'write',
+            'files': [{
+                'path': '/foo/bar',
+                'make-dirs': True,
+                'permissions': '600',
+                'user': 'bob',
+                'group': 'staff',
+            }],
+        })
+
+    def test_write_file_path_error(self):
+        self.client.responses.append((
+            {'Content-Type': 'application/json'},
+            b"""
+{
+    "result": [
+        {"path": "/foo/bar", "error": {"kind": "not-found", "message": "not found"}}
+    ],
+    "status": "OK",
+    "status-code": 200,
+    "type": "sync"
+}
+""",
+        ))
+
+        source = io.BytesIO(b'content')
+        with self.assertRaises(pebble.PathError) as cm:
+            self.client.write_file('/foo/bar', source)
+        self.assertEqual(cm.exception.kind, 'not-found')
+        self.assertEqual(cm.exception.message, 'not found')
+
+        self.assertEqual(len(self.client.requests), 1)
+        request = self.client.requests[0]
+        self.assertEqual(request[:3], ('POST', '/v1/files', None))
+
+        headers, body = request[3:]
+        content_type = headers['Content-Type']
+        dest = io.BytesIO()
+        req = self._parse_write_multipart(content_type, body, {'/foo/bar': dest})
+        self.assertEqual(dest.getvalue(), b'content')
+        self.assertEqual(req, {
+            'action': 'write',
+            'files': [{'path': '/foo/bar'}],
+        })
+
+    def _parse_write_multipart(self, content_type, body, destinations):
```

Comment:
it seems odd to have a straight copy of this function here. Is it reasonable to have a separate _private/multipart.py file and import it from both places?

Reply:
It was somewhat different, and I didn't want to have to pollute the main code with test-specific options to try to unify it. In any case, I've removed this function in pebble.py now with the new read_content implementation.